### PR TITLE
Allow removing unreferenced composable data stream templates

### LINKED ISSUE #84171: Index Template cannot be deleted / error message seems incorrect
### Elasticsearch Version

7.17

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Not relevant

### Problem Description

Creating a match all index template is bad and we know it.
I needed it for a test and I created a new index template with a `*` `index_pattern`.

I wanted to delete it after few minutes (so that it wouldn't picked up by any new index creation) and I couldn't, unless updating it to match no index.

It seems to deny the deletion because it reports being used by data streams.

### Steps to Reproduce

Create a Index Template matching all:
```
PUT _index_template/matchall
{
  "index_patterns": [
    "*"
  ],
  "priority": 1,
  "template": {
    "mappings": {
      "_meta": {
        "class": "lol"
      }
    }
  }
}
```

Try to delete it:
```
DELETE _index_template/matchall
```

You get:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "illegal_argument_exception",
        "reason" : "unable to remove composable templates [matchall] as they are in use by a data streams [.fleet-actions-results, .logs-deprecation.elasticsearch-default, .slm-history-5, ilm-history-5, logs-elastic_agent-default, logs-elastic_agent.endpoint_security-default, logs-elastic_agent.filebeat-default, logs-elastic_agent.metricbeat-default, logs-endpoint.events.file-default, logs-endpoint.events.library-default, logs-endpoint.events.network-default, logs-endpoint.events.process-default, logs-endpoint.events.registry-default, logs-endpoint.events.security-default, logs-generic-default, logs-windows.powershell-default, logs-windows.powershell_operational-default, logs-windows.sysmon_operational-default, metrics-elastic_agent.elastic_agent-default, metrics-elastic_agent.endpoint_security-default, metrics-elastic_agent.filebeat-default, metrics-elastic_agent.metricbeat-default, metrics-endpoint.metadata-default, metrics-endpoint.metrics-default, metrics-endpoint.policy-default, metrics-system.cpu-default, metrics-system.diskio-default, metrics-system.memory-default, metrics-system.network-default, metrics-system.process-default, metrics-system.process_summary-default, metrics-system.socket_summary-default, metrics-system.uptime-default, metrics-windows.perfmon-default, metrics-windows.service-default]"
      }
    ],
    "type" : "illegal_argument_exception",
    "reason" : "unable to remove composable templates [matchall] as they are in use by a data streams [.fleet-actions-results, .logs-deprecation.elasticsearch-default, .slm-history-5, ilm-history-5, logs-elastic_agent-default, logs-elastic_agent.endpoint_security-default, logs-elastic_agent.filebeat-default, logs-elastic_agent.metricbeat-default, logs-endpoint.events.file-default, logs-endpoint.events.library-default, logs-endpoint.events.network-default, logs-endpoint.events.process-default, logs-endpoint.events.registry-default, logs-endpoint.events.security-default, logs-generic-default, logs-windows.powershell-default, logs-windows.powershell_operational-default, logs-windows.sysmon_operational-default, metrics-elastic_agent.elastic_agent-default, metrics-elastic_agent.endpoint_security-default, metrics-elastic_agent.filebeat-default, metrics-elastic_agent.metricbeat-default, metrics-endpoint.metadata-default, metrics-endpoint.metrics-default, metrics-endpoint.policy-default, metrics-system.cpu-default, metrics-system.diskio-default, metrics-system.memory-default, metrics-system.network-default, metrics-system.process-default, metrics-system.process_summary-default, metrics-system.socket_summary-default, metrics-system.uptime-default, metrics-windows.perfmon-default, metrics-windows.service-default]"
  },
  "status" : 400
}
```

The list seems to be composed by all the data streams.

## Workaround

Update the `index_patterns` to something non-existing.
```
PUT _index_template/matchall
{
  "index_patterns": [
    "none"
  ],
  "priority": 1,
  "template": {
    "mappings": {
      "_meta": {
        "class": "lol"
      }
    }
  }
}
```

Then delete it:
```
DELETE _index_template/matchall
```

### Logs (if relevant)

_No response_

### LINKED ISSUE #84188: Lower priority index template that matches data stream cannot be removed
### Elasticsearch Version

7.x, 8.x

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Darwin 

### Problem Description

A data stream must have an index template that matches it in order to exist.
When managing index templates we want to make sure we don't leave a
data stream without a [corresponding index template](https://github.com/elastic/elasticsearch/blob/acf9968f9d397d755d5ad9fcf815cca23bbeda61/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java#L918).

The `dataStreamsUsingTemplate` check uses the index pattern of the index 
templates to conclude if a template is used by a data stream, however it should
be enhanced to check the priority too as there can only be only one index template
that matches a data stream and it's the one with the highest priority.

### Steps to Reproduce

Create an ILM policy
```
PUT _ilm/policy/testing_yo
{
  "policy": {
    "phases": {
      "warm": {
	"actions": {
	  "shrink": {
          "number_of_shards": 1
	  }
        }
      }
    } 
  }
}
```

Create an index template with the `data_stream` option:
```
PUT _index_template/logs_data_stream
{
	"index_patterns": [ "logs*", "te*", "bar*" ],
	"data_stream": {},
        "priority": 100,
	"template" : {
	  "settings": {
            "index.number_of_shards": 3,
	    "index.lifecycle.name": "testing"
	  }
	}
}
```

Create the data stream
```
POST logs_test/_doc 
{
  "message": "testing",
  "@timestamp": "1591890611"
}
```

Create a new index template that would match this data stream with a higher priority
```
PUT _index_template/higher_logs_data_stream
{
	"index_patterns": [ "logs*", "te*", "bar*" ],
	"data_stream": {},
        "priority": 500,
	"template" : {
	  "settings": {
            "index.number_of_shards": 3,
	    "index.lifecycle.name": "testing"
	  }
	}
}
```

Attempt to delete the previous, lower priority index template:
```
DELETE _index_template/logs_data_stream
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by dakrone**
Previously there was a bug where when we retrieved the list of data streams using a template, we
only checked index patterns, rather than actually comparing priorities and only treating the highest
priority template as "referenced". This commit fixes this problem by only treating the matching
composable template as referenced, allowing unreferenced templates to be safely deleted.

Resolves #84188
Resolves #84171

### CODE REVIEW

**Review by andreidan — APPROVED** [review]
LGTM, thanks for fixing this Lee

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java [hunk 1]
```diff
@@ -663,6 +664,25 @@ public void testCannotDeleteComposableTemplateUsedByDataStream() throws Exceptio
...
+                null,
+                null,
+                new ComposableIndexTemplate.DataStreamTemplate(),
+                null
+            )
+        );
+        client().execute(PutComposableIndexTemplateAction.INSTANCE, request).actionGet();
+
+        DeleteComposableIndexTemplateAction.Request deleteRequest = new DeleteComposableIndexTemplateAction.Request("id");
+        client().execute(DeleteComposableIndexTemplateAction.INSTANCE, deleteRequest).get();
```
[the remark this answers is not in the document]
Comment (dakrone) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/DataStreamIT.java:686]: Good idea, I've added this.

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
@elasticmachine update branch

**Pull request merged**